### PR TITLE
node-lifecycle-controller: improve monitorNodeHealth test-coverage

### PR DESCRIPTION
**Pull request by atosatto**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

With the recent removal of the `--enable-taint-manager` flag in https://github.com/kubernetes/kubernetes/pull/115840, NodeLifecycleController is not anymore performing Pod evictions, in order to better reflect the actual implementation of the controller, the tests covering the `monitorNodeHealth` function have been updated in this PR to ensure the state of the zone is set correctly in case of changes to the Node status.

#### Which issue(s) this PR fixes:

Follow-up from https://github.com/kubernetes/kubernetes/pull/115840

#### Special notes for your reviewer:

The tests have been updated to remove the test-cases covering Pod eviction as these do not apply anymore.

The coverage has now been increased to cover 64.1% of the statements vs the current 60.3% of statements coverage

##### Current test coverage

```console
$ make test WHAT=./pkg/controller/nodelifecycle KUBE_COVER=y
go version go1.20.1 darwin/arm64
WARNING: ulimit -n (files) should be at least 1000, is 256, may cause test failure
+++ [0407 17:42:29] Running tests with code coverage and with -race
+++ [0407 17:42:29] Saving coverage output in '/tmp/k8s_coverage/20230407-174229'
skipped	k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/generator
skipped	k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/rest
ok  	k8s.io/kubernetes/pkg/controller/nodelifecycle	8.769s	coverage: 60.3% of statements
+++ [0407 17:42:40] Combined coverage report: /tmp/k8s_coverage/20230407-174229/combined-coverage.html
```

##### New test coverage

```console
$ make test WHAT=./pkg/controller/nodelifecycle KUBE_COVER=y
go version go1.20.2 darwin/arm64
WARNING: ulimit -n (files) should be at least 1000, is 256, may cause test failure
+++ [0407 17:41:22] Running tests with code coverage and with -race
+++ [0407 17:41:22] Saving coverage output in '/tmp/k8s_coverage/20230407-174122'
skipped	k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/generator
skipped	k8s.io/kubernetes/vendor/k8s.io/client-go/1.4/rest
ok  	k8s.io/kubernetes/pkg/controller/nodelifecycle	8.924s	coverage: 64.1% of statements
+++ [0407 17:41:34] Combined coverage report: /tmp/k8s_coverage/20230407-174122/combined-coverage.html
```

##### Detailed diff of test coverage

The `monitorNodeHealth` function coverage improved from `50.8%` to `78.8%`.

```diff
$ diff <(go tool cover -func=/tmp/k8s_coverage/20230407-174229/combined-coverage.out) <(go tool cover -func=/tmp/k8s_coverage/20230407-174122/combined-coverage.out)
1c1
< k8s.io/kubernetes/pkg/controller/nodelifecycle/metrics.go:98:				Register				0.0%
---
> k8s.io/kubernetes/pkg/controller/nodelifecycle/metrics.go:98:				Register				100.0%
7,19c7,19
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:295:	NewNodeLifecycleController		26.3%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:459:	Run					27.3%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:512:	doNodeProcessingPassWorker		31.6%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:536:	doNoScheduleTaintingPass		0.0%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:591:	doNoExecuteTaintingPass			32.1%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:661:	monitorNodeHealth			50.8%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:772:	processTaintBaseEviction		18.8%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:812:	isNodeExcludedFromDisruptionChecks	75.0%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:821:	tryUpdateNodeHealth			78.0%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:987:	handleDisruption			93.8%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1078:	podUpdated				100.0%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1088:	doPodProcessingWorker			20.0%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1106:	processPod				95.2%
---
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:295:	NewNodeLifecycleController		39.0%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:459:	Run					0.0%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:512:	doNodeProcessingPassWorker		0.0%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:536:	doNoScheduleTaintingPass		52.0%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:591:	doNoExecuteTaintingPass			83.8%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:661:	monitorNodeHealth			78.8%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:772:	processTaintBaseEviction		75.0%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:812:	isNodeExcludedFromDisruptionChecks	100.0%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:821:	tryUpdateNodeHealth			92.3%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:987:	handleDisruption			85.7%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1078:	podUpdated				0.0%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1088:	doPodProcessingWorker			0.0%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1106:	processPod				0.0%
21,23c21,23
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1169:	classifyNodes				87.5%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1201:	HealthyQPSFunc				100.0%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1207:	ReducedQPSFunc				100.0%
---
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1169:	classifyNodes				46.7%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1201:	HealthyQPSFunc				0.0%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1207:	ReducedQPSFunc				0.0%
25,26c25,26
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1230:	markNodeForTainting			63.6%
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1248:	markNodeAsReachable			100.0%
---
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1230:	markNodeForTainting			100.0%
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1248:	markNodeAsReachable			66.7%
28,29c28,29
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1292:	reconcileNodeLabels			0.0%
< total:											(statements)				57.3%
---
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1292:	reconcileNodeLabels			81.8%
> total:											(statements)				64.1%
```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

None.

### CODE REVIEW

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 1]
```diff
@@ -167,561 +166,685 @@ func newNodeLifecycleControllerFromClient(
...
+					t.Errorf("unexpected initial zone state: %v: got %v instead of %v", zone, nodeController.zoneStates[zone], state)
+				}
 			}
-		}
 
-		if item.expectedEvictPods != podEvicted {
-			t.Errorf("expected pod eviction: %+v, got %+v for %+v", item.expectedEvictPods,
-				podEvicted, item.description)
-		}
+			// following zone state
```
Comment (Huang-Wei) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:828]: nit: do you want to use a for loop to de-dupe code of this section with previous "initial zone state" section?

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 2]
```diff
@@ -167,561 +166,685 @@ func newNodeLifecycleControllerFromClient(
...
-					{
-						ObjectMeta: metav1.ObjectMeta{
-							Name:              "node0",
-							CreationTimestamp: fakeNow,
+		"No Disruption: Initially both zones down, one comes back": {
+			nodeList: []*v1.Node{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:              "node0",
+						CreationTimestamp: metav1.Date(2012, 1, 1, 0, 0, 0, 0, time.UTC),
```
Comment (Huang-Wei) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:234]: do you want to unify all usage of "metav1.Date(2012, 1, 1, 0, 0, 0, 0, time.UTC)"-like to base on `fakeNow`?

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 3]
```diff
@@ -167,561 +166,685 @@ func newNodeLifecycleControllerFromClient(
...
-						Status: v1.NodeStatus{
-							Conditions: []v1.NodeCondition{
-								{
-									Type:               v1.NodeReady,
-									Status:             v1.ConditionFalse,
-									LastHeartbeatTime:  metav1.Date(2015, 1, 1, 12, 0, 0, 0, time.UTC),
-									LastTransitionTime: metav1.Date(2015, 1, 1, 12, 0, 0, 0, time.UTC),
-								},
-							},
+		"Partial Disruption: one Node failed leading to the number of healthy Nodes to exceed the configured threshold": {
```
Comment (Huang-Wei) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:355]: qq: is the default threshold 50%?

Comment (ravisantoshgudimetla) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:355]: it is computed dynamically based on `nc.unhealthyZoneThreshold`

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go (hunk 1 above)
Comment (ravisantoshgudimetla) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:828]: or a method.

**Review by ravisantoshgudimetla — APPROVED** [review]
Thank you for following up on this @atosatto. Just some nits and clarifications.

/lgtm

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 4]
```diff
@@ -167,561 +166,685 @@ func newNodeLifecycleControllerFromClient(
...
+			if err := nodeController.monitorNodeHealth(ctx); err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+			if len(nodeController.zoneStates) != len(tt.expectedInitialStates) {
+				t.Errorf("unexpected length of initial zone state map: got %d instead of %d", len(nodeController.zoneStates), len(tt.expectedInitialStates))
+			}
+			for zone, state := range tt.expectedInitialStates {
+				if state != nodeController.zoneStates[zone] {
+					t.Errorf("unexpected initial zone state: %v: got %v instead of %v", zone, nodeController.zoneStates[zone], state)
+				}
```
[the remark this answers is not in the document]
Comment (atosatto) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: Thanks, fixed in bb5a3bcd7e8 to use `gocmp`

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go (hunk 2 above)
Comment (atosatto) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: Yeah good point, changed this in bb5a3bcd7e8

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go (hunk 3 above)
Comment (atosatto) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:356]: All of the tests are using the `testUnhealthyThreshold` constant which is set to `0.55`.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 5]
```diff
@@ -1200,7 +1202,7 @@ func (nc *Controller) HealthyQPSFunc(nodeNum int) float32 {
 	return nc.evictionLimiterQPS
 }
 
-// ReducedQPSFunc returns the QPS for when a the cluster is large make
+// ReducedQPSFunc returns the QPS for when the cluster is large make
```
[the remark this answers is not in the document]
Comment (atosatto) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: Yeah good point, reverted this change to reduce the scope

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 6]
```diff
@@ -653,9 +653,11 @@ func (nc *Controller) doNoExecuteTaintingPass(ctx context.Context) {
 	}
 }
 
-// monitorNodeHealth verifies node health are constantly updated by kubelet, and
-// if not, post "NodeReady==ConditionUnknown".
-// This function will taint nodes who are not ready or not reachable for a long period of time.
+// monitorNodeHealth verifies node health are constantly updated by kubelet, and if not, post "NodeReady==ConditionUnknown".
+// This function will
+//   - add nodes who are not ready or not reachable for a long period of time to a rate-limited
```
[the remark this answers is not in the document]
Comment (atosatto) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: Changed in bb5a3bcd7e8, thanks!

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 7]
```diff
@@ -167,561 +166,685 @@ func newNodeLifecycleControllerFromClient(
...
-	labels := map[string]string{
-		v1.LabelTopologyRegion:          "region1",
-		v1.LabelTopologyZone:            "zone1",
-		v1.LabelFailureDomainBetaRegion: "region1",
-		v1.LabelFailureDomainBetaZone:   "zone1",
-	}
-
-	// Because of the logic that prevents NC from evicting anything when all Nodes are NotReady
-	// we need second healthy node in tests. Because of how the tests are written we need to update
-	// the status of this Node.
```
[the remark this answers is not in the document]
Comment (atosatto) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:181]: I don't think these comments are relevant anymore as we are not performing any eviction. What we are doing in this function is validating the state transition between the initial and following zone state.

**Review by Huang-Wei — COMMENTED** [review]
/lgtm

Thanks @atosatto!

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go (hunk 1 above)
Comment (atosatto) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:824]: Just noticed that I missed this review from earlier commits - changed in a17b1aecbec

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 8]
```diff
@@ -653,9 +653,11 @@ func (nc *Controller) doNoExecuteTaintingPass(ctx context.Context) {
 	}
 }
 
-// monitorNodeHealth verifies node health are constantly updated by kubelet, and
-// if not, post "NodeReady==ConditionUnknown".
-// This function will taint nodes who are not ready or not reachable for a long period of time.
+// monitorNodeHealth verifies node health are constantly updated by kubelet, and if not, post "NodeReady==ConditionUnknown".
```
Comment (soltysh) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go:656]: Nit: 
```suggestion
// monitorNodeHealth verifies if node health are constantly updated by kubelet, and if not, post "NodeReady==ConditionUnknown".
```

**Review by soltysh — APPROVED** [review]
/approve

**Pull request merged**